src/index/git2.rs

```diff
@@ -49,28 +49,21 @@ where
 {
     let config = repo.config()?;
 
-    let mut tried_sshkey = false;
-    let mut tried_cred_helper = false;
-    let mut tried_default = false;
-
     f(&mut |url, username, allowed| {
         if allowed.contains(git2::CredentialType::USERNAME) {
             return Err(git2::Error::from_str("no username specified in remote URL"));
         }
 
-        if allowed.contains(git2::CredentialType::SSH_KEY) && !tried_sshkey {
-            tried_sshkey = true;
+        if allowed.contains(git2::CredentialType::SSH_KEY) {
             let username = username.unwrap();
             return git2::Cred::ssh_key_from_agent(username);
         }
 
-        if allowed.contains(git2::CredentialType::USER_PASS_PLAINTEXT) && !tried_cred_helper {
-            tried_cred_helper = true;
+        if allowed.contains(git2::CredentialType::USER_PASS_PLAINTEXT) {
             return git2::Cred::credential_helper(&config, url, username);
         }
 
-        if allowed.contains(git2::CredentialType::DEFAULT) && !tried_default {
-            tried_default = true;
+        if allowed.contains(git2::CredentialType::DEFAULT) {
             return git2::Cred::default();
         }
```

Comment:
hi there, i had a look at the approach you linked in Cargo. I can see that they check the flags down the line after they're set.
I'm not too familiar with git2 either, but i made these changes after looking at the code semantics. I'll explain my reasoning, and maybe you can see where i'm coming from (or point out where I'm going wrong!)

- you first set all the flags to false
- then in the first 'if' block you have `!tried_ssh_key`, which always evaluates to `true` since the flag has just been set to `false`. at this point you either set it to `true`, or leave it alone, depending on the value of `allowed.contains(git2::CredentialType::SSH_KEY)`. either way, the value of the flag never gets read again.
- the same is true for the next 2 blocks. The value is deterministically set when you make the comparison, which makes the comparison redundent, and then never read again afterwards.

am i missing something?

Reply:
In the first run, yes, the check `!tried_ssh_key` will always pass, and it could be set to true (depending on `allowed.contains(git2::CredentialType::SSH_KEY)`) and same for the two other branches.  
But in the next runs of the callback (`git2` might call it again if our first credentials didn't succeed), the flag (that we have set to `true` in the previous run) will cause its branch to no longer be taken and fallthrough to the next authentication method instead.  
In this scheme, `allowed` is basically contains what the user wants to try authenticating with, so it allows to opt-out from some authentication methods but doesn't change during the entire process (at least, I don't think so), so we rely on our own flags to track progress through the different authentication methods.  
In Cargo's case, their flags are `tried_sshkey` and `cred_helper_bad`, and they also use it afterwards because they try even further with different SSH usernames if it has been attempted (like `git` or `${USER}`), which we do not do right now.

Reply:
aha i see. did not occur to me that the callback would be run multiple times, thanks for explanation!
i'll drop that commit